### PR TITLE
Fix name resolution in typedef and allow defaults for template tags

### LINKED ISSUE #45480: Incorrect name resolution for JSDoc template tag constraints in typedef
# Bug Report

### 🔎 Search Terms

jsdoc name resolve lookup symbol template constraint typedef

### 🕗 Version & Regression Information

- This has been an issue since we added support for `@template` in JSDoc

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&jsx=0&filetype=js#code/PTAEAEGMAsFNINYCkDOAuUAXATgV1gLABQIEAhgDYUD2A7qhjvsaeCjgJaSaN6FEsAVIOKhBETLAC2ABwplJoACqjx4SbPmKA3glgBPagDNlAX1ABpVRP0zYAE1gmAYtWqrgxIA)

### 💻 Code

```js
/**
 * @template T
 * @template {keyof T} K
 * @typedef Foo
 */
```

### 🙁 Actual behavior

Reports `` Cannot find name `T` ``.

### 🙂 Expected behavior

Does not report an error.

### Notes

Name resolution does not look at the symbol table of the `@typedef` tag, so when looking for `T` in the code below, we end up looking at `T` for the function declaration further down in this example:

```js
/**
 * @template T
 * @template {keyof T} K
 * @typedef Foo
 */

/**
 * @template T
 * @param {T} a
 */
function bar(a) {
}
```

A "go-to-definition" on the `T` in `keyof T` jumps to the `T` declared for `bar`, not the `T` in the typedef for `Foo`.

### LINKED ISSUE #45481: Support type parameter defaults in JSDoc template tags
# Suggestion

Add support for type parameter defaults in JSDoc `@template` tags, to match parity with type parameter defaults in TypeScript:

```ts
// ts
type Foo<T = number> = T

// js
/**
 * @template [T=number]
 * @typedef {T} Foo
 */
```

## 🔍 Search Terms

jsdoc template type parameter default 

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Implement type parameter defaults in `@template` similar to optional parameters in `@param`:

```js
/**
 * @param {string} [x=1]  - NOTE: TS parses the expression `1` but ignores it.
 */
function f(x) {}

/**
 * @template {string} [T="hello"]
 * @typedef {T} Foo
 */
```

- Brackets around `T` would be required (so as not to misinterpret existing documentation comments
- A bracketed template type *must* have a default (unlike `@param` which allows `[name]` on its own).

## 📃 Motivating Example

Improves JavaScript JSDoc type annotation parity with TypeScript type annotations.

## 💻 Use Cases

Used for JavaScript development using the TypeScript language service.

### PR DESCRIPTION
**Pull request by rbuckton**
This adds support for Type Parameter defaults in JavaScript files using the JSDoc `@template` tag. The implementation borrows from the syntax used for parameter defaults in JSDoc (namely `[foo=1]`):

```js
/**
 * @template {string} [T="hello"]
 * @typedef {T} Foo
 */
```

This also fixes a name resolution bug when checking the constraint of a `@template` tag.

Fixes #45480
Fixes #45481

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -8053,6 +8053,22 @@ namespace ts {
...
+                function parseBracketNameInTemplateTag(): { name: Identifier, defaultType?: TypeNode } {
+                    const isBracketed = parseOptionalJsdoc(SyntaxKind.OpenBracketToken);
+                    if (isBracketed) {
+                        skipWhitespace();
+                    }
+                    const name = parseJSDocIdentifierName(Diagnostics.Unexpected_token_A_type_parameter_name_was_expected_without_curly_braces);
+                    let defaultType: TypeNode | undefined;
+                    if (isBracketed) {
+                        skipWhitespace();
+                        parseExpected(SyntaxKind.EqualsToken);
```
Comment (sandersn) [inline src/compiler/parser.ts:8065]: Silly question: what would just `[T]` mean?

Probably nothing, but I was just thinking about the way the parser allows almost any combination of characters for existing JSDoc syntax.

**Review by sandersn — CHANGES_REQUESTED** [review]
Looks good, one request in the parser and one in the tests.

File: tests/baselines/reference/jsdocTemplateTagDefault.errors.txt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+tests/cases/conformance/jsdoc/file.js(33,14): error TS2706: Required type parameters may not follow optional type parameters.
+tests/cases/conformance/jsdoc/file.js(39,14): error TS2706: Required type parameters may not follow optional type parameters.
+tests/cases/conformance/jsdoc/file.js(44,17): error TS2744: Type parameter defaults can only reference previously declared type parameters.
+tests/cases/conformance/jsdoc/file.js(59,14): error TS2706: Required type parameters may not follow optional type parameters.
+tests/cases/conformance/jsdoc/file.js(66,17): error TS2744: Type parameter defaults can only reference previously declared type parameters.
+
+
+==== tests/cases/conformance/jsdoc/file.js (8 errors) ====
+    /**
+     * @template {string | number} [T=string] - ok: defaults are permitted
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply tests/baselines/reference/jsdocTemplateTagDefault.errors.txt:13]: > `/** @template [A,K=string,V=number,B] */`

Invalid syntax

> `/** @template A,[K=string],B,[V=number],C */`

Valid syntax

> `/** @template {string | number} [K=string,V=number] */`

Invalid syntax, but this works:

> `/** @template {string | number} [K=string], [V=number] */`

We have a NOTE that we should disallow constraints and commas, but we don't enforce that. We could disallow `,` with bracketed names if we wanted.

File: src/compiler/parser.ts (hunk 1 above)
Comment (rbuckton) [reply src/compiler/parser.ts:8065]: `[T]` is a syntax error, and just means `T`. For the bracketed syntax to work for a `@template` tag you *must* specify `=type`

**Comment by rbuckton** [comment]
I'm debating whether `[T]` should be allowed, and if so, if it should default to the constraint or `unknown`. Given the changes necessary to support that, I may look at that in a separate PR instead.

**Comment by rbuckton** [comment]
@sandersn can you take another look?

**Comment by rbuckton** [comment]
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
- User Tests: lots of diffs in baselines, but none seem related to this change.
  - NOTE: Prettier has same OOM as https://github.com/microsoft/TypeScript/pull/45350#issuecomment-905914392, so I think there's something unrelated to this PR or #45350.
- RWC: no changes of note.

**Comment by andrewbranch** [comment]
@rbuckton FYI you can run `@typescript-bot user test this inline` to get a user test diff between this and current main, with results posted in the PR.

**Comment by rbuckton** [comment]
Wasn't aware of the `inline`, and I'm not sure why that's not the default behavior, tbh.

**Review by sandersn — APPROVED** [review]

**Comment by rbuckton** [comment]
@DanielRosenwasser: Do we want to discuss the syntax for this in a design meeting, or is this ok to merge?

File: tests/cases/conformance/jsdoc/jsdocTemplateTagNameResolution.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// @allowJs: true
+// @checkJs: true
+// @noEmit: true
+// @Filename: file.js
```
Comment (weswigham) [inline tests/cases/conformance/jsdoc/jsdocTemplateTagNameResolution.ts:4]: Could you also enable declaration emit for this (or both) test(s)? JS declaration emit is pretty specialized, after all.

**Comment by rbuckton** [comment]
TODO:
- [ ] Verify JS declaration emit.

**Review by weswigham — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
So the syntax

```
@template {string} [T="hello"]
```

means

```
T extends string = "hello"
```

I can't wait to never get this right. 😂

**Comment by thw0rted** [comment]
Hi @rbuckton , I noticed that none of the changes in this PR seem to include any documentation.  Is that hosted in a different repo?  If not, is there a task / issue to track inclusion of this syntax in the handbook?

**Comment by orta** [comment]
Just noting that the new JSDoc reference on the site includes this

**Pull request merged**